relay_hub/storage.py

```
@@ -0,0 +1,627 @@
+            created_at, last_id = self._decode_cursor(request.cursor)
+            filters.append("(created_at < ? OR (created_at = ? AND id < ?))")
+            cursor_time = self._format_dt(created_at)
+            params.extend([cursor_time, cursor_time, last_id])
+
+        where_clause = " AND ".join(filters) if filters else "1=1"
+
+        if request.query:
+            sql = (
+                "SELECT m.* FROM messages m JOIN messages_fts fts ON m.rowid = fts.rowid "
+                f"WHERE (fts.body MATCH ? OR fts.subject MATCH ? OR fts.tags MATCH ?) AND {where_clause} "
+                "ORDER BY m.created_at DESC, m.id DESC LIMIT ?"
```

Ruff (0.13.1)

Possible SQL injection vector through string-based query construction

(S608)